### PR TITLE
New implicit top drag feature

### PR DESCRIPTION
**Pull request by cbegeman**
This PR creates a new option to use implicit top drag in the same manner as implicit bottom drag. It assumes that the only setting in which top drag will be used is in land ice-covered regions. 

Implicit top drag is activated by setting `config_use_implicit_top_drag = .true.` and `config_implicit_top_drag_coeff` > 0 and `config_land_ice_flux_mode = 'standalone'` (while also ensuring that at least part of the domain has `landIceFraction` > 0).

[BFB]

### CODE REVIEW

**Comment by cbegeman** [comment]
E3SM-Ocean-Discussion needs to be updated to reflect v2.1

**Comment by cbegeman** [comment]
The use of implicit top drag instead of explicit top drag with the same drag coefficient (2.5e-3) results in a stronger overturning and barotropic streamfunction and friction velocity.

Explicit:
image
image
image


Implicit:
image
image
image

**Comment by cbegeman** [comment]
This branch is BFB with master-v2.1 across the nightly compass test suite on chrysalis (intel, openmpi).

**Comment by cbegeman** [comment]
When merged, I will open a compass PR to update the isomip_plus namelist. In the meantime, the update is here https://github.com/cbegeman/compass/tree/ocn-new-impl-top-drag-feature.

**Comment by xylar** [comment]
Those are very interesting results.  I'll review as soon as I can.

**Review by xylar — APPROVED** [review]
This looks excellent and I think it's ready for E3SM.

File: components/mpas-ocean/src/shared/mpas_ocn_vmix.F [hunk 1]
```diff
@@ -463,15 +467,18 @@ subroutine ocn_vel_vmix_tend_implicit(dt, kineticEnergyCell, vertViscTopOfEdge,
...
 
            ! tridiagonal matrix algorithm
            C(Nsurf)     = -2.0_RKIND*dt*vertViscTopOfEdge(Nsurf+1,iEdge) &
                       / (layerThickEdgeMean(Nsurf,iEdge) + layerThickEdgeMean(Nsurf+1,iEdge)) &
                       / layerThickEdgeMean(Nsurf,iEdge)
-           bTemp(Nsurf) = 1.0_RKIND - C(Nsurf)
+           bTemp(Nsurf) = 1.0_RKIND - C(Nsurf) &
+              + dt * implicitTopDragCoef * landIceEdgeFraction(iEdge) &
+                * sqrt(kineticEnergyCell(N,cell1) + kineticEnergyCell(N,cell2)) / &
+                layerThickEdgeMean(N,iEdge)
```
Comment (xylar) [inline components/mpas-ocean/src/shared/mpas_ocn_vmix.F:481]: Hmm, one question. Shouldn't the `N`s here be `Nsurf`s?
```suggestion
                * sqrt(kineticEnergyCell(Nsurf,cell1) + kineticEnergyCell(Nsurf,cell2)) / &
                layerThickEdgeMean(Nsurf,iEdge)
```

**Comment by xylar** [comment]
I noticed a potential issue, and it might possibly explain the differences between explicit and implicit drag.  It could also be that I've misunderstood the code.

File: components/mpas-ocean/src/shared/mpas_ocn_vmix.F (hunk 1 above)
Comment (cbegeman) [reply components/mpas-ocean/src/shared/mpas_ocn_vmix.F:481]: @xylar I think you're right. Good catch! I'll test this out today.

**Comment by cbegeman** [comment]
After the fix https://github.com/E3SM-Ocean-Discussion/E3SM/pull/37/commits/1aa8323a1c800e245913851d05a90051bb0feec3, the fields are similar to baseline:
image
image
image

Diffs are small:
```temperature          Time index: 0
0:  l1: 4.94494964689862e-06  l2: 3.54671497803758e-07  linf: 1.11879863373687e-07 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/output.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/output.nc

salinity             Time index: 0
0:  l1: 1.52896312499706e-06  l2: 1.09559425618205e-07  linf: 3.49838273905334e-08 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/output.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/output.nc

layerThickness       Time index: 0
0:  l1: 9.28252126852769e-06  l2: 1.53320768783754e-07  linf: 8.20610202012517e-09 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/output.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/output.nc

normalVelocity       Time index: 0
0:  l1: 1.00379330431482e-04  l2: 3.60327612492113e-06  linf: 9.40049915661110e-07 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/output.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/output.nc

ssh                  Time index: 0
0:  l1: 9.28251652609902e-06  l2: 4.84841767732003e-07  linf: 8.19766228232766e-08 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIcePressure      Time index: 0
  PASS /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceDraft         Time index: 0
  PASS /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceFraction      Time index: 0
  PASS /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceMask          Time index: 0
  PASS /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceFrictionVelocity Time index: 0
0:  l1: 4.38181454439973e-08  l2: 6.39115327570759e-09  linf: 2.46878993648020e-09 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

topDrag              Time index: 0
0:  l1: 9.12076846010558e-08  l2: 8.63644997345552e-09  linf: 2.90341464757310e-09 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

topDragMagnitude     Time index: 0
0:  l1: 4.32268530739718e-08  l2: 6.57047118884458e-09  linf: 2.55111726218826e-09 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceFreshwaterFlux Time index: 0
0:  l1: 1.04167968684903e-08  l2: 1.07060544661501e-09  linf: 2.16226914113516e-10 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceHeatFlux      Time index: 0
0:  l1: 3.54021066873278e-03  l2: 3.63737137560911e-04  linf: 7.34271149980259e-05 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

heatFluxToLandIce    Time index: 0
0:  l1: 3.22392987069271e-05  l2: 3.26062719949709e-06  linf: 6.39598907170580e-07 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceBoundaryLayerTemperature Time index: 0
0:  l1: 1.33529659999921e-06  l2: 1.80206523145301e-07  linf: 6.62909558357683e-08 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceBoundaryLayerSalinity Time index: 0
0:  l1: 4.00354707608130e-07  l2: 5.44320543335968e-08  linf: 1.99685032953312e-08 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceHeatTransferVelocity Time index: 0
0:  l1: 8.50072021629417e-10  l2: 1.23988373549072e-10  linf: 4.78945247685562e-11 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceSaltTransferVelocity Time index: 0
0:  l1: 2.42877718579886e-11  l2: 3.54252493116635e-12  linf: 1.36841498281453e-12 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceInterfaceTemperature Time index: 0
0:  l1: 2.89274246556559e-07  l2: 4.48166356518168e-08  linf: 1.64615954201253e-08 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

landIceInterfaceSalinity Time index: 0
0:  l1: 5.11089547927668e-06  l2: 7.92284744921169e-07  linf: 2.91053368073335e-07 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

accumulatedLandIceMass Time index: 0
0:  l1: 3.12042131308428e-05  l2: 3.60431876046923e-06  linf: 1.41614001467061e-06 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

accumulatedLandIceHeat Time index: 0
0:  l1: 1.00271847844368e-01  l2: 1.17968345443340e-02  linf: 4.73071687497395e-03 
  FAIL /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/ocn/new-impl-top-drag-feature-v2/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc

       /lcrc/group/e3sm/ac.cbegeman/scratch/MPAS-Ocean-output/chrys/master-v2.1/ocean/isomip_plus/5km/sigma/Ocean0/performance/land_ice_fluxes.nc
```

**Review by xylar — APPROVED** [review]
Perfect, I'm very comfortable with the differences you're now seeing.  Not machine round-off but still pretty small as we would expect.

This is ready to move to E3SM.

**Comment by cbegeman** [comment]
Thanks @xylar!

This PR has been moved to https://github.com/E3SM-Project/E3SM/pull/5447

**Pull request closed without merging**